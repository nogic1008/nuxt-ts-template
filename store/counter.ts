import { Module, mutation, VuexModule } from 'vuex-class-component'

export interface Counter {
  count: number
  increment(): void
  decrement(): void
}

@Module({ namespacedPath: 'counter', target: 'nuxt' })
export class CounterStore extends VuexModule implements Counter {
  count = 0

  @mutation increment() {
    this.count++
  }

  @mutation decrement() {
    this.count--
  }
}
