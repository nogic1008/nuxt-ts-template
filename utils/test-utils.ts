/** Generate [0-9a-z]{length} string. */
export const generateRandomString = (length: number) =>
  [...Array(Number.isInteger(length) ? length : 12)]
    .map(() => (~~(Math.random() * 36)).toString(36))
    .join('')
