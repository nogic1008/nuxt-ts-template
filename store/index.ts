import Vuex from 'vuex'
import { Counter, CounterStore } from '~/store/counter'

export const store = new Vuex.Store({
  modules: {
    counter: CounterStore.ExtractVuexModule(CounterStore)
  },
  strict: false
})

export interface VuexTypeModule {
  counter: Counter
}

export const vxm: VuexTypeModule = {
  counter: CounterStore.CreateProxy(store, CounterStore)
}
