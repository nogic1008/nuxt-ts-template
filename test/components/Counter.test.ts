import { shallowMount, createLocalVue, Wrapper, mount } from '@vue/test-utils'
import Buefy from 'buefy'
import { VuexTypeModule } from '~/store'
import CounterComponent from '~/components/Counter.vue'
import { CounterStore } from '~/store/counter'

const localVue = createLocalVue()
localVue.use(Buefy)

describe('components/Counter.vue', () => {
  let wrapper: Wrapper<CounterComponent>
  let $vxm: VuexTypeModule

  beforeEach(() => {
    $vxm = {
      counter: new CounterStore()
    }
    wrapper = shallowMount(CounterComponent, { localVue, mocks: { $vxm } })
  })

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  test('renders correctly', () => {
    const wrapper = mount(CounterComponent, { localVue, mocks: { $vxm } })
    expect(wrapper.element).toMatchSnapshot()
  })

  test.each([0, 1, 2, 10])(
    'click plus button increases count',
    async (pressCount) => {
      // Arrange
      const wrapper = mount(CounterComponent, { localVue, mocks: { $vxm } })
      const button = wrapper.find('button.plus')

      // Act
      for (let i = 0; i < pressCount; i++) {
        button.trigger('click')
      }

      // Assert
      await wrapper.vm.$nextTick()
      expect(wrapper.find('span.subtitle').text()).toBe(pressCount.toString())
    }
  )

  test.each([0, 1, 2, 10])(
    'click minus button decreases count',
    async (pressCount) => {
      // Arrange
      const wrapper = mount(CounterComponent, { localVue, mocks: { $vxm } })
      const button = wrapper.find('button.minus')

      // Act
      for (let i = 0; i < pressCount; i++) {
        button.trigger('click')
      }

      // Assert
      await wrapper.vm.$nextTick()
      expect(wrapper.find('span.subtitle').text()).toBe(
        (-pressCount).toString()
      )
    }
  )
})
