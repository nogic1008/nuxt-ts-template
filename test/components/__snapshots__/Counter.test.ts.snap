// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Counter.vue renders correctly 1`] = `
<div
  class="column"
>
  <div
    class="card"
  >
    <header
      class="card-header"
    >
      <p
        class="card-header-title has-text-grey"
      >
        
        Store Sample
      
      </p>
    </header>
     
    <div
      class="card-content"
    >
      <div
        class="content has-text-centered"
      >
        <span
          class="icon has-text-primary is-large"
        >
          <i
            class="mdi mdi-counter mdi-48px"
          />
        </span>
      </div>
    </div>
     
    <footer
      class="card-footer"
    >
      <div
        class="card-footer-item"
      >
        <span>
          <button
            class="button plus"
            type="button"
          >
            <span
              class="icon is-small"
            >
              <i
                class="mdi mdi-plus"
              />
            </span>
             
            <!---->
             
            <!---->
          </button>
           
          <span
            class="subtitle"
          >
            0
          </span>
           
          <button
            class="button minus"
            type="button"
          >
            <span
              class="icon is-small"
            >
              <i
                class="mdi mdi-minus"
              />
            </span>
             
            <!---->
             
            <!---->
          </button>
        </span>
      </div>
    </footer>
  </div>
</div>
`;
